design/ErrorHandling.md

```diff
@@ -23,59 +23,140 @@ There are several types of WebGPU calls that get their errors handled differentl
 Implementations should provide a way to enable synchronous validation, for example via a debug shim or via the developer tools.
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
-## *Telemetry* \& *Fallback*: Error Logging
+The behavior is implementation-defined. For example, it may throw an exception, or it may just cause the debugger to pause inside the call that fails.
 
-`GPUDevice` keeps a log of errors that happened on each thread (main thread or Web Worker).
-The application "eventually" recieves these errors, asynchronously via a callback.
+## *Fatal Errors*: Lost/Recovered Events
 
-These errors are all delivered through one stream - not directly associated with the objects or operations that produced them.
-They should not be used by applications to recover from expected, recoverable errors.
-Instead, the error log may be used for handling unexpected errors in deployment, for bug reporting and telemetry.
+TODO: distinguish device and adapter loss
+
+The `GPUAdapter` and `GPUDevice` are event targets which receive events about device and adapter status.
+
+
+TODO!
+TODO!
+TODO!: redo text with these 3 events
 
+```webidl
+partial interface GPUAdapter : EventTarget {
+    readonly attribute boolean isReady;
+};
+
+interface GPUAdapterLostEvent : Event {
+    readonly attribute DOMString reason;
+};
+
+interface GPUAdapterReadyEvent : Event {};
 ```
-enum GPULogEntryType {
-    "device-lost",
-    "validation-error",
-    "recoverable-out-of-memory",
+
+```webidl
+partial interface GPUDevice : EventTarget {};
+
+interface GPUDeviceLostEvent : Event {
+    readonly attribute boolean recoverable;
+    readonly attribute DOMString reason;
 };
+```
+
+ - `GPUAdapter` `"gpuadapterlost" -> GPUAdapterLostEvent`:
+ - `GPUAdapter` `"gpuadapterready" -> GPUAdapterReadyEvent`:
+   Signals when it is okay to create new devices on this adapter.
+   It may fire only if:
+    - the adapter is still valid, and
+    - the adapter's `isReady` attribute was previously `false`.
+
+ - `GPUDevice` `"gpudevicelost" -> GPUDeviceLostEvent`:
+   Signals that the `GPUDevice` cannot be used anymore.
+   Any further operations on the device, or on objects associated with the device, are errors.
+   (The `"gpulogentry"` event will not fire after a device loss, so this makes all further operations on the device effectively no-ops.)
+   This may happen if something goes fatally wrong on the device (e.g. unexpected out-of-memory, crash, or native device loss).
+   This event **may** cause the adapter's `isReady` attribute to become `false`.
+   The application must check `isReady` before attempting to create a new device.
+
+   TODO::::::::::::::::::::::::::: The application **may not** attempt creation of a new device on this adapter until it receives the `"gpuadapterready"` event.
+
+TODO: How does an app detect when createDevice gave them a lost device
```

Comment:
The event probably fires immediately.

Reply:
For the moment, this is a callback, but it might be possible to have the event fire immediately if we choose to use an event instead.